simulation/model.py

```diff
@@ -84,6 +90,34 @@ def increment_timestep(self):
             self.boiler_upgrade_scheme_spend_gbp
         )
 
+    def get_heat_pump_price_discount_schedule(
+        self,
+        heat_pump_price_discount_schedule: Optional[
+            List[Tuple[datetime.datetime, float]]
+        ],
+    ):
+
+        if heat_pump_price_discount_schedule:
+
+            discount_schedule = {
+                schedule[0]: schedule[1]
+                for schedule in heat_pump_price_discount_schedule
+            }
+
+            price_change_dates = discount_schedule.keys()
+            first_date, last_date = min(price_change_dates), max(price_change_dates)
```

Comment:
You don't need to call `.keys()` on the dictionary - `max` and `min` operate on the keys anyway, for example:

```
>>> x = {'a': 3, 'b': 2, 'c': 1}
>>> max(x)
'c'
>>> min(x)
'a'
```

So you can just use `min(discount_schedule), max(discount_schedule)`.